### PR TITLE
Support non-square pixels

### LINKED ISSUE #13573: COG trouble with non-square pixels
**Describe the bug**
I'm loading a COG, but only the upper part is shown.

**To Reproduce**
1. Go to https://codesandbox.io/s/cog-forked-x0mxoc?file=/main.js (this is the cog example from openlayers just with a different URL)

**Expected behavior**
It should show completely, like in e.g. geotiff.io or QGIS:
image

### PR DESCRIPTION
**Pull request by tschaub**
This adds support for tile grids with non-uniform resolution (non-square pixels).

The `resolution` of a tile (or the view) corresponds with the `xResolution`.

The `aspectRatio` of the tile is the ratio `xResolution / yResolution`.

So a tile that is 100px wide and 50px tall with a `resolution` of 1 and an `aspectRatio` of 1/2 will cover 100 x 100 map units.

The GeoTIFF source now supports tiles with non-square pixels.  The WebGL tile renderer can be used to render such tiles.

Fixes #13573.

### CODE REVIEW

File: src/ol/tilegrid/TileGrid.js [hunk 1]
```diff
@@ -40,7 +40,11 @@ const DECIMALS = 5;
  * `origin` must be provided.
  * @property {!Array<number>} resolutions Resolutions. The array index of each resolution needs
  * to match the zoom level. This means that even if a `minZoom` is configured, the resolutions
- * array will have a length of `maxZoom + 1`.
+ * array will have a length of `maxZoom + 1`.  If the grid is configured with an apsect ratios array,
+ * the resolutions array should correpsond with the x-resolution.
```
Comment (marcjansen) [inline src/ol/tilegrid/TileGrid.js:44]:
```suggestion
 * the resolutions array should correspond with the x-resolution.
```

File: src/ol/tilegrid/TileGrid.js [hunk 2]
```diff
@@ -40,7 +40,11 @@ const DECIMALS = 5;
  * `origin` must be provided.
  * @property {!Array<number>} resolutions Resolutions. The array index of each resolution needs
  * to match the zoom level. This means that even if a `minZoom` is configured, the resolutions
- * array will have a length of `maxZoom + 1`.
+ * array will have a length of `maxZoom + 1`.  If the grid is configured with an apsect ratios array,
```
Comment (marcjansen) [inline src/ol/tilegrid/TileGrid.js:43]:
```suggestion
 * array will have a length of `maxZoom + 1`.  If the grid is configured with an aspect ratios array,
```

File: src/ol/tilegrid/TileGrid.js (hunk 1 above)
Comment (tschaub) [reply src/ol/tilegrid/TileGrid.js]: Thanks for the catch!

File: src/ol/tilegrid/TileGrid.js (hunk 2 above)
Comment (tschaub) [reply src/ol/tilegrid/TileGrid.js]: Corrected in the latest commit.

**Comment by mike-000** [comment]
If the texture size does not match the specified tile size that can be already handled using tilePixelRatio.  Could this be achieved more simply by using a 2 dimensional tilePixelRatio (and if tilePixelRatio changes with z level adding a tilePixelRatio array to the tile grid) and handling that in TileTexture, avoiding changes to TileGrid and renderer functionality?  In your implementation will tile grids be compatible between renderers - e.g. if you use the tile grid from a GeoTIFF source in a DebugTile source will it give the same result in WebGL and canvas Tile layers?

**Comment by mike-000** [comment]
If the texture size does not match the specified tile size that can be already handled using tilePixelRatio.  Could this be achieved more simply by using a 2 dimensional tilePixelRatio (and if tilePixelRatio changes with z level adding a tilePixelRatio array to the tile grid) and handling that in TileTexture, avoiding changes to TileGrid and renderer functionality?  In your implementation will tile grids be compatible between renderers - e.g. if you use the tile grid from a GeoTIFF source in a DebugTile source will it give the same result in WebGL and canvas Tile layers?

**Comment by tschaub** [comment]
There is not a mismatch between texture size and tile size in the case of non-square pixels.  For example, a GeoTIFF may have 256x256 tiles, represented by typed arrays of 256x256xNBANDS pixels.  The pixels are “non-square” in the sense that their x- and y-resolutions do not match.

**Comment by tschaub** [comment]
And no, tile grids with aspect ratio other than 1 are not supported in the canvas renderer.  This means you cannot render a GeoTIFF source in both a canvas and webgl tile layer.  This is already the case before this change, so not an additional limitation on valid use cases.

**Comment by mike-000** [comment]
> There is not a mismatch between texture size and tile size in the case of non-square pixels. For example, a GeoTIFF may have 256x256 tiles, represented by typed arrays of 256x256xNBANDS pixels. The pixels are “non-square” in the sense that their x- and y-resolutions do not match.

In your test case you have textures of 150 x 75 in a conventional tile size of 150 x 150.  That could be implemented as an existing [150, 150] tile grid with 2 dimensional tilePixelRatio of [1, 0.5].

**Comment by mike-000** [comment]
> And no, tile grids with aspect ratio other than 1 are not supported in the canvas renderer. This means you cannot render a GeoTIFF source in both a canvas and webgl tile layer. This is already the case before this change, so not an additional limitation on valid use cases.

Aspect ratio other than 1 are supported in the canvas renderer - if images do not fit they are scaled without needing to specify the aspect ratio or tilePixelRatio, see #13605.

**Comment by tschaub** [comment]
> Aspect ratio other than 1 are supported in the canvas renderer

The term "aspect ratio" in this pull request refers to this ratio: `xResolution / yResolution`.  Where the `xResolution` and `yResolution` are the number of map units per pixel in the x and y directions.

You may be confusing this with the ratio `tilePixelWidth / tilePixelHeight`.  The two are not the same.

**Comment by tschaub** [comment]
@mike-000 - A new scalar `pixelAspectRatio` property for sources could work.  Then the GeoTIFF source would then create a tile grid not using the actual tile width and height but instead one modified by the pixel aspect ratio.  The shortcoming of this approach is that this would only support a single pixel aspect ratio for all levels.

The current solution of using an array of aspect ratios allows for this value to vary per zoom level.  The x- and y-resolutions of overviews in the GeoTIFF may vary (though there is no telling how common this is).

We could also configure sources with an array of pixel aspect ratios.  This would need to match the length of levels in the tile grid.

So the alternatives are
 1. configure the source with a new property that modifies the interpretation of the tile grid (specifically the tile sizes), or
 2. configure the tile grid with a new property

Alternative 1 is a smaller code change (and the tile grid sizes are "wrong").  Alternative 2 is a larger code change.

**Comment by tschaub** [comment]
@mike-000 - I updated the rendering test to make it clearer what this change is about.  The test uses a GeoTIFF with non-square pixels.  Here is a sandbox with the data included: https://codesandbox.io/s/data-tiles-forked-f0b66f

**Comment by mike-000** [comment]
> @mike-000 - I updated the rendering test to make it clearer what this change is about. The test uses a GeoTIFF with non-square pixels. Here is a sandbox with the data included: https://codesandbox.io/s/data-tiles-forked-f0b66f. The example in that sandbox doesn't work in part because the tile sizes in the tile grid are the actual tile sizes, not modified by the pixel aspect ratio.

It works as expected if I get tiles from the GeoTIFF source and then use the data in a DataTile source https://codesandbox.io/s/data-tiles-forked-5wb78r?file=/main.js  The 256 x 128 tiles (as understood by other sources/renderers) contain 256 x 256 textures.  The GeoTIFF source seems unable to get some of those tiles for itself.

**Comment by mike-000** [comment]
> We could also configure sources with an array of pixel aspect ratios. This would need to match the length of levels in the tile grid.

Yes, I would consider pixel aspect ratio to be a property of the source, just a using retina tiles is.  The tile grid should not need to know about it.  We already allow `size` or `sizes`, `url` or `urls`, so both single or array could be allowed.

**Comment by mike-000** [comment]
It may be better to always scale up to see the detail.  If the aspect ratio is > 1 base the tile size on the y texture size, if < 1 on the x texture size https://codesandbox.io/s/data-tiles-forked-1s2z7d?file=/main.js

**Comment by tschaub** [comment]
This branch attempted to reinterpret the tile grid's tile size as "source" tile size.  The current interpretation throughout most of the library is that the tile grid's tile size is the "rendered" tile size.  I didn't like how we sometimes use `tilePixelRatio` and sometimes use `image.width` and `image.height` to make decisions about scaling source tiles to rendered tiles, so I wanted to make things more explicit for the webgl renderer.  But this feels like it goes against the current interpretation of tile sizes.  I now think a better approach is to give data tiles an explicit source size (see #13648).

**Pull request closed without merging**